util/math_reference.cpp

```diff
@@ -310,6 +310,15 @@ T mul_hi_unsigned(T x, T y) {
   return hi;
 }
 
+/**
+ * @brief Function to get high sizeof(T)*8 bits of the product of two signed T
```

Comment:
I guess the fact that `T` is an integer is missing. Can `T` be signed or unsigned?

Reply:
Add some `static_assert(std::is_integral_v...)`?

Reply:
`mul_hi_signed` is only available inside math_reference.cpp and is only called below in `mul_hi` with signed input parameters, so it seems to me that we can do without static type checking.